### PR TITLE
Add jsdoc member names: Class#method

### LINKED ISSUE #43594: @link and @see should support Class#method syntax
Extends https://github.com/microsoft/TypeScript/pull/43312
Based on the analysis in https://github.com/microsoft/TypeScript/issues/35524#issuecomment-686698291

### PR DESCRIPTION
**Pull request by sandersn**
This PR adds support for jsdoc instance references of the form `Class#method`. Right now, these are parsed only inside of `@see` and `@link` tags, but they could probably be allowed other places fairly simply.

## Additional fixes

1. Allow find-all-refs (and related services like goto-def) on all parts of an entity name. Previously in `@see {A.B.c}`, only `c` worked. Now all 3 do.
2. Put missing space into `@link` quick info for links with unresolved entities. I think I was going to follow up with a general fix that proved too ambitious and forgot the ad-hoc fix until now.

## Implementation notes

1. The parser and scanner now have a `rescanPrivateIdentifier`, which breaks up `#id` into two tokens, `# id`. This is the main way to get the new `HashToken` out of the scanner, but `scanJsDocToken` now also returns `HashToken` instead of `Unknown`. My first attempt just reused PrivateIdentifier, but this gave bad spans for find-all-refs. I also think that always scanning `#` as part of a private identifier ties that token too tightly to private identifiers, so when `#` gets used for other things, it'll probably make sense to invert the scanner's behaviour and provide a `rescanHash` function when private identifiers are desired.
2. `resolveJSDocInstanceReference` is recursive, then delegates to `resolveEntityName` whenever it encounters a normal EntityName like `K.x` in, for example, `K.x#m`. However, when `K.x` doesn't resolve, it *also* tries `K.prototype.x`. That's because both `K#x` and `K.x` are supported inside `@see/@link` as synonyms of `K.prototype.x`. This fallback is inefficient, but only happens in services when find-all-refs is requested for `@see/@link`, and only performs redundant work for chained entity names like `A.B.c` which should actually be `A#B#c`. 

Fixes #43594

### CODE REVIEW

**Comment by sandersn** [comment]
@DanielRosenwasser @amcasey I added you because you've had good opinions about usability in the past. Feel free to skip the implementation.

@andrewbranch @elibarzilay I heard you like parsing....

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -7305,7 +7312,13 @@ namespace ts {
             export function parseJSDocNameReference(): JSDocNameReference {
                 const pos = getNodePos();
                 const hasBrace = parseOptional(SyntaxKind.OpenBraceToken);
-                const entityName = parseEntityName(/* allowReservedWords*/ false);
+                const p2 = getNodePos();
+                let entityName: EntityName | JSDocInstanceReference = parseEntityName(/* allowReservedWords*/ false);
+                while (token() === SyntaxKind.PrivateIdentifier) {
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts:7317]: `while`?

So are you supporting `Foo#Bar#Baz`?

Comment (sandersn) [reply src/compiler/parser.ts:7317]: A left-deep JSDocInstanceReference tree: `((Foo#Bar)#Baz)`

File: src/compiler/scanner.ts [hunk 2]
```diff
@@ -2251,6 +2254,14 @@ namespace ts {
             return token;
         }
 
+        function reScanPrivateIdentifier(): SyntaxKind {
```
Comment (andrewbranch) [inline src/compiler/scanner.ts:2257]: I know the various `reScan` functions don’t all have perfectly comparable behavior, but it seems like the general naming convention is naming it for what you _want_, not what you _currently have_: “I currently have token `X`, but have reason to believe that could be interpreted as token `Y`, so to re-scan `X` as `Y`, I call `reScanY`.”

The naming convention is clearly bad when read like that, but diverging from it is probably worse. I would either conform, naming this `reScanHashToken`, or be totally explicit, naming it `reScanPrivateIdentifierAsHashToken`.

**Review by andrewbranch — CHANGES_REQUESTED** [review]

File: src/compiler/scanner.ts (hunk 2 above)
Comment (sandersn) [reply src/compiler/scanner.ts:2257]: Nope, I just confused myself. I will reverse it.

File: src/compiler/parser.ts [hunk 3]
```diff
@@ -7760,9 +7773,17 @@ namespace ts {
                     nextTokenJSDoc(); // start at token after link, then skip any whitespace
                     skipWhitespace();
                     // parseEntityName logs an error for non-identifier, so create a MissingNode ourselves to avoid the error
-                    const name = tokenIsIdentifierOrKeyword(token())
+                    const p2 = getNodePos();
+                    let name: EntityName | JSDocInstanceReference | undefined = tokenIsIdentifierOrKeyword(token())
                         ? parseEntityName(/*allowReservedWords*/ true)
                         : undefined;
+                    if (name) {
+                        while(token() === SyntaxKind.PrivateIdentifier) {
```
[the remark this answers is not in the document]
Comment (sandersn) [reply src/compiler/parser.ts:7781]: Well, it *used* to. I'm not sure why it didn't complain this time.

**Review by andrewbranch — APPROVED** [review]

**Pull request merged**